### PR TITLE
The `..` isn't the "spread operator"

### PR DESCRIPTION
**Pull request by BillWagner**
Refer to `..e` as a spread element instead.

### CODE REVIEW

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by jnm2** [comment]
@BillWagner Why would we prefer to avoid calling the `..` part of the `..e` spread expression a spread operator?

**Comment by RikkiGibson** [comment]
FWIW, at some point we felt it was more clear to call `e` in `..e` the "spread operand", rather than the `spread expression` or `spread element expression`, for example.

It feels reasonable to me to name the parts as follows:
- `..e` in `[..e]` is a spread element.
- `..` in `[..e]` is a spread operator.
- `e` in `[..e]` is a spread operand.

**Comment by BillWagner** [comment]
@jnm2 I think it makes the description simpler.  I'll defer to both @CyrusNajmabadi and @MadsTorgersen for a more complete answer.

**Comment by CyrusNajmabadi** [comment]
Honestly, we're just trying to have simple and consistent naming everywhere.  We want to think of this as just different sorts of elements.  A simple expression element.  A spread element.  And, coming up, a key-value-pair element.

Bringing the concepts of 'operators' into the mix just adds more conceptual overhead.

**Comment by jnm2** [comment]
I think it's natural for someone to want to refer to `..` specifically. "Add a spread operator, then it will compile." That's simpler than saying "turn the simple element expression into a spread expression."

**Comment by CyrusNajmabadi** [comment]
I'm ok with however people want to talk informally.  But i'd like the spec to talk about this consistently.  For example, i don't want to have to update the operators section, or have to discuss things like 'operator precedence' here with this.

**Comment by RikkiGibson** [comment]
Could you please clarify what are the formal names of the parts enumerated in https://github.com/dotnet/csharplang/pull/8122#issuecomment-2110887662?

**Comment by CyrusNajmabadi** [comment]
> Could you please clarify what are the formal names of the parts enumerated in https://github.com/dotnet/csharplang/pull/8122#issuecomment-2110887662?

1. `..e` in `[..e]` is a spread element.
1. `..` in [..e] is "the `..` token"
1. `e` in `[..e]` is the expression to spread

**Comment by RikkiGibson** [comment]
I would encourage to include in the spec that those specific names refer to those specific parts. Then, when we want to refer to these parts without ambiguity, we can use those terms.

To emphasize my point: I think it is a worthwhile goal for spec authoring to adopt terms which are distinct and terse enough that the need for informal terms alongside them is minimized. Informal terms, especially in use among the authors/collaborators/implementers of a spec, increase ambiguity, and can sometimes make discussing/interpreting the feature and its spec harder, not easier.

**Comment by tannergooding** [comment]
> For example, i don't want to have to update the operators section, or have to discuss things like 'operator precedence' here with this.

I really think we should do that anyways.

In general, people are going to talk about and have questions around what `..` means and how to refer to it (in the context of both ranges/slicing and spread). So even if the language wants to explicitly talk about `..` as a token, that's the perfect reason to add an informative note which acknowledges that many people and other languages may consider it an `operator`, but that in C# `..` is a `token` and does not play into operator precedence. Such an informative note or annotation belongs in places that people may look at and expect to find `..` (such as `Operator Overloading/Precedence`)

-- Personally I think it makes more sense to just call `..` an operator. It's clearly called such for `Range` which is used as part of slice expressions, so it's potentially already confusing that `[..e]` sometimes means `spread token` and other times `range operator` depending on the surrounding context of where `[..e]` appears.

**Comment by CyrusNajmabadi** [comment]
I think it will be more confusing if we make it an operator here as well.  I also would want us to update the impl to match that design.

**Comment by CyrusNajmabadi** [comment]
I don't think changing the naming helps with confusion.  Using your example:  we don't call `[...]` the "collection operator" even though `[...]` is called the "indexer operator" when you have `a[0]`. 

Similarly `?` is the conditional operator when it's an operator. But we don't call it an operator when it is used to specify a nullable type.

Etc. etc 

Just because we use the same tokens doesn't mean we make them the same way.

**Comment by jnm2** [comment]
What makes a thing an operator in a user mental model?

**Comment by CyrusNajmabadi** [comment]
To me, loosely, an operator needs to take an expression and operate on it (and possibly more expressions) to *produce* a new expression. The latter part is important. It's a way of composing up expressions to create more expressions. 

For .. to be an operator, I would want it to apply to an expression (it does) and to produce something that is an expression.  I would also expect it to compose upwards. 


For example, say it was `...` (to avoid range ambiguity), if it was an operator, I'd expect to be able to do things like `[x ? ...y : ...z]`.  

It would really be something that fits and composes and produces expressions uniformly like all the rest of the operators do. 

But it really doesn't :-). It's just a marker on an element, only in one location, only for collection elements. 

Basically, an 'operator' comes with a lot of meaning that doesn't apply here.  That's my 2c take :-)

**Comment by CyrusNajmabadi** [comment]
nope  there's no such thing (at hte lang levle) of a spread-element-expression.  Note: i would expect *any* expression to be embeddable somehow in other experessions.  Thsi would be the only expression that didn't work that way if we made this an expression.
